File: release/com/example/budgettingapp/database/BudgetDatabaseHelper.java
package com.example.budgettingapp.database;

import android.content.ContentValues;
import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.provider.BaseColumns;
import com.example.budgettingapp.data.Expense;
import com.example.budgettingapp.data.Goal;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

@kotlin.Metadata(mv = {1, 9, 0}, k = 1, xi = 48, d1 = {"\u0000`\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010\u000e\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010\b\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010 \n\u0002\b\u0003\n\u0002\u0010\t\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u0006\n\u0002\b\u0005\n\u0002\u0010\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\n\u0018\u0000 *2\u00020\u0001:\u0004)*+,B\r\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u00a2\u0006\u0002\u0010\u0004J\u000e\u0010\u000b\u001a\u00020\f2\u0006\u0010\r\u001a\u00020\u000eJ\u000e\u0010\u000f\u001a\u00020\f2\u0006\u0010\u0010\u001a\u00020\u0011J\f\u0010\u0012\u001a\b\u0012\u0004\u0012\u00020\u000e0\u0013J\f\u0010\u0012\u001a\b\u0012\u0004\u0012\u00020\u000e0\u0013J\f\u0010\u0014\u001a\b\u0012\u0004\u0012\u00020\u00110\u0013J\u0010\u0010\u0015\u001a\u0004\u0018\u00010\u000e2\u0006\u0010\u0016\u001a\u00020\u0017J\f\u0010\u0018\u001a\b\u0012\u0004\u0012\u00020\u00190\u0013J\u0016\u0010\u001a\u001a\u00020\u001b2\u0006\u0010\u001c\u001a\u00020\u00062\u0006\u0010\u001d\u001a\u00020\u0006J\u000e\u0010\u001e\u001a\u00020\u00172\u0006\u0010\r\u001a\u00020\u000eJ\u000e\u0010\u001f\u001a\u00020\u00172\u0006\u0010\u0010\u001a\u00020\u0011J\u0010\u0010 \u001a\u00020!2\u0006\u0010\"\u001a\u00020#H\u0016J \u0010$\u001a\u00020!2\u0006\u0010\"\u001a\u00020#2\u0006\u0010%\u001a\u00020\f2\u0006\u0010&\u001a\u00020\fH\u0016J\u000e\u0010\'\u001a\u00020\f2\u0006\u0010\r\u001a\u00020\u000eJ\u000e\u0010(\u001a\u00020\f2\u0006\u0010\u0010\u001a\u00020\u0011R\u000e\u0010\u0005\u001a\u00020\u0006X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0007\u001a\u00020\u0006X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u0016\u0010\b\u001a\n \n*\u0004\u0018\u00010\t0\tX\u0082\u0004\u00a2\u0006\u0002\n\u0000\u00a8\u0006-"}, d2 = {"Lcom/example/budgettingapp/database/BudgetDatabaseHelper;", "Landroid/database/sqlite/SQLiteOpenHelper;", "context", "Landroid/content/Context;", "(Landroid/content/Context;)V", "SQL_CREATE_ENTRIES", "", "SQL_DELETE_ENTRIES", "dateFormatter", "Ljava/time/format/DateTimeFormatter;", "kotlin.jvm.PlatformType", "deleteExpense", "", "expense", "Lcom/example/budgettingapp/data/Expense;", "deleteGoal", "goal", "Lcom/example/budgettingapp/data/Goal;", "getAllExpenses", "", "getAllGoals", "getExpenseById", "id", "", "getExpensesByCategory", "Lcom/example/budgettingapp/database/BudgetDatabaseHelper$CategoryTotal;", "getTotalExpenses", "", "startDate", "endDate", "insertExpense", "insertGoal", "onCreate", "", "db", "Landroid/database/sqlite/SQLiteDatabase;", "onUpgrade", "oldVersion", "newVersion", "updateExpense", "updateGoal", "CategoryTotal", "Companion", "ExpenseEntry", "GoalEntry", "app_release"})
public final class BudgetDatabaseHelper extends android.database.sqlite.SQLiteOpenHelper {
    @org.jetbrains.annotations.NotNull
    private final java.lang.String SQL_CREATE_ENTRIES = null;
    @org.jetbrains.annotations.NotNull
    private final java.lang.String SQL_DELETE_ENTRIES = null;
    @org.jetbrains.annotations.NotNull
    public static final java.lang.String DATABASE_NAME = "budget_database.db";
    public static final int DATABASE_VERSION = 2;
    private final java.time.format.DateTimeFormatter dateFormatter = null;
    @org.jetbrains.annotations.NotNull
    public static final com.example.budgettingapp.database.BudgetDatabaseHelper.Companion Companion = null;
    
    public BudgetDatabaseHelper(@org.jetbrains.annotations.NotNull
    android.content.Context context) {
        super(null, null, null, 0);
    }
    
    @java.lang.Override
    public void onCreate(@org.jetbrains.annotations.NotNull
    android.database.sqlite.SQLiteDatabase db) {
    }
    
    @java.lang.Override
    public void onUpgrade(@org.jetbrains.annotations.NotNull
    android.database.sqlite.SQLiteDatabase db, int oldVersion, int newVersion) {
    }
    
    public final long insertExpense(@org.jetbrains.annotations.NotNull
    com.example.budgettingapp.data.Expense expense) {
        return 0L;
    }
    
    public final int updateExpense(@org.jetbrains.annotations.NotNull
    com.example.budgettingapp.data.Expense expense) {
        return 0;
    }
    
    @org.jetbrains.annotations.NotNull
    public final java.util.List<com.example.budgettingapp.data.Expense> getAllExpenses() {
        return null;
    }
    
    public final long insertGoal(@org.jetbrains.annotations.NotNull
    com.example.budgettingapp.data.Goal goal) {
        return 0L;
    }
    
    public final int updateGoal(@org.jetbrains.annotations.NotNull
    com.example.budgettingapp.data.Goal goal) {
        return 0;
    }
    
    @org.jetbrains.annotations.NotNull
    public final java.util.List<com.example.budgettingapp.data.Goal> getAllGoals() {
        return null;
    }
    
    public final int deleteGoal(@org.jetbrains.annotations.NotNull
    com.example.budgettingapp.data.Goal goal) {
        return 0;
    }
    
    public final int deleteExpense(@org.jetbrains.annotations.NotNull
    com.example.budgettingapp.data.Expense expense) {
        return 0;
    }
    
    @org.jetbrains.annotations.NotNull
    public final java.util.List<com.example.budgettingapp.data.Expense> getAllExpenses() {
        return null;
    }
    
    @org.jetbrains.annotations.Nullable
    public final com.example.budgettingapp.data.Expense getExpenseById(long id) {
        return null;
    }
    
    public final double getTotalExpenses(@org.jetbrains.annotations.NotNull
    java.lang.String startDate, @org.jetbrains.annotations.NotNull
    java.lang.String endDate) {
        return 0.0;
    }
    
    @org.jetbrains.annotations.NotNull
    public final java.util.List<com.example.budgettingapp.database.BudgetDatabaseHelper.CategoryTotal> getExpensesByCategory() {
        return null;
    }
    
    @kotlin.Metadata(mv = {1, 9, 0}, k = 1, xi = 48, d1 = {"\u0000(\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0000\n\u0002\u0010\u000e\n\u0000\n\u0002\u0010\u0006\n\u0002\b\t\n\u0002\u0010\u000b\n\u0002\b\u0002\n\u0002\u0010\b\n\u0002\b\u0002\b\u0086\b\u0018\u00002\u00020\u0001B\u0015\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u00a2\u0006\u0002\u0010\u0006J\t\u0010\u000b\u001a\u00020\u0003H\u00c6\u0003J\t\u0010\f\u001a\u00020\u0005H\u00c6\u0003J\u001d\u0010\r\u001a\u00020\u00002\b\b\u0002\u0010\u0002\u001a\u00020\u00032\b\b\u0002\u0010\u0004\u001a\u00020\u0005H\u00c6\u0001J\u0013\u0010\u000e\u001a\u00020\u000f2\b\u0010\u0010\u001a\u0004\u0018\u00010\u0001H\u00d6\u0003J\t\u0010\u0011\u001a\u00020\u0012H\u00d6\u0001J\t\u0010\u0013\u001a\u00020\u0003H\u00d6\u0001R\u0011\u0010\u0002\u001a\u00020\u0003\u00a2\u0006\b\n\u0000\u001a\u0004\b\u0007\u0010\bR\u0011\u0010\u0004\u001a\u00020\u0005\u00a2\u0006\b\n\u0000\u001a\u0004\b\t\u0010\n\u00a8\u0006\u0014"}, d2 = {"Lcom/example/budgettingapp/database/BudgetDatabaseHelper$CategoryTotal;", "", "category", "", "total", "", "(Ljava/lang/String;D)V", "getCategory", "()Ljava/lang/String;", "getTotal", "()D", "component1", "component2", "copy", "equals", "", "other", "hashCode", "", "toString", "app_release"})
    public static final class CategoryTotal {
        @org.jetbrains.annotations.NotNull
        private final java.lang.String category = null;
        private final double total = 0.0;
        
        public CategoryTotal(@org.jetbrains.annotations.NotNull
        java.lang.String category, double total) {
            super();
        }
        
        @org.jetbrains.annotations.NotNull
        public final java.lang.String getCategory() {
            return null;
        }
        
        public final double getTotal() {
            return 0.0;
        }
        
        @org.jetbrains.annotations.NotNull
        public final java.lang.String component1() {
            return null;
        }
        
        public final double component2() {
            return 0.0;
        }
        
        @org.jetbrains.annotations.NotNull
        public final com.example.budgettingapp.database.BudgetDatabaseHelper.CategoryTotal copy(@org.jetbrains.annotations.NotNull
        java.lang.String category, double total) {
            return null;
        }
        
        @java.lang.Override
        public boolean equals(@org.jetbrains.annotations.Nullable
        java.lang.Object other) {
            return false;
        }
        
        @java.lang.Override
        public int hashCode() {
            return 0;
        }
        
        @java.lang.Override
        @org.jetbrains.annotations.NotNull
        public java.lang.String toString() {
            return null;
        }
    }
    
    @kotlin.Metadata(mv = {1, 9, 0}, k = 1, xi = 48, d1 = {"\u0000\u0018\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0002\b\u0002\n\u0002\u0010\u000e\n\u0000\n\u0002\u0010\b\n\u0000\b\u0086\u0003\u0018\u00002\u00020\u0001B\u0007\b\u0002\u00a2\u0006\u0002\u0010\u0002R\u000e\u0010\u0003\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0005\u001a\u00020\u0006X\u0086T\u00a2\u0006\u0002\n\u0000\u00a8\u0006\u0007"}, d2 = {"Lcom/example/budgettingapp/database/BudgetDatabaseHelper$Companion;", "", "()V", "DATABASE_NAME", "", "DATABASE_VERSION", "", "app_release"})
    public static final class Companion {
        
        private Companion() {
            super();
        }
    }
    
    @kotlin.Metadata(mv = {1, 9, 0}, k = 1, xi = 48, d1 = {"\u0000\u0014\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010\u000e\n\u0002\b\t\b\u00c6\u0002\u0018\u00002\u00020\u0001B\u0007\b\u0002\u00a2\u0006\u0002\u0010\u0002R\u000e\u0010\u0003\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0005\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0006\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0007\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\b\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\t\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\n\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u000b\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\f\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000\u00a8\u0006\r"}, d2 = {"Lcom/example/budgettingapp/database/BudgetDatabaseHelper$ExpenseEntry;", "Landroid/provider/BaseColumns;", "()V", "COLUMN_AMOUNT", "", "COLUMN_CATEGORY", "COLUMN_DATE", "COLUMN_DESCRIPTION", "COLUMN_END_TIME", "COLUMN_START_TIME", "COLUMN_TIMESTAMP", "COLUMN_TITLE", "TABLE_NAME", "app_release"})
    public static final class ExpenseEntry implements android.provider.BaseColumns {
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String TABLE_NAME = "expenses";
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String COLUMN_TITLE = "title";
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String COLUMN_AMOUNT = "amount";
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String COLUMN_CATEGORY = "category";
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String COLUMN_DATE = "date";
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String COLUMN_START_TIME = "start_time";
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String COLUMN_END_TIME = "end_time";
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String COLUMN_DESCRIPTION = "description";
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String COLUMN_TIMESTAMP = "timestamp";
        @org.jetbrains.annotations.NotNull
        public static final com.example.budgettingapp.database.BudgetDatabaseHelper.ExpenseEntry INSTANCE = null;
        
        private ExpenseEntry() {
            super();
        }
    }
    
    @kotlin.Metadata(mv = {1, 9, 0}, k = 1, xi = 48, d1 = {"\u0000\u0014\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010\u000e\n\u0002\b\u0007\b\u00c6\u0002\u0018\u00002\u00020\u0001B\u0007\b\u0002\u00a2\u0006\u0002\u0010\u0002R\u000e\u0010\u0003\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0005\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0006\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0007\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\b\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\t\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\n\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000\u00a8\u0006\u000b"}, d2 = {"Lcom/example/budgettingapp/database/BudgetDatabaseHelper$GoalEntry;", "Landroid/provider/BaseColumns;", "()V", "COLUMN_CURRENT_AMOUNT", "", "COLUMN_DEADLINE", "COLUMN_DESCRIPTION", "COLUMN_TARGET_AMOUNT", "COLUMN_TIMESTAMP", "COLUMN_TITLE", "TABLE_NAME", "app_release"})
    public static final class GoalEntry implements android.provider.BaseColumns {
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String TABLE_NAME = "goals";
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String COLUMN_TITLE = "title";
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String COLUMN_TARGET_AMOUNT = "target_amount";
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String COLUMN_CURRENT_AMOUNT = "current_amount";
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String COLUMN_DEADLINE = "deadline";
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String COLUMN_DESCRIPTION = "description";
        @org.jetbrains.annotations.NotNull
        public static final java.lang.String COLUMN_TIMESTAMP = "timestamp";
        @org.jetbrains.annotations.NotNull
        public static final com.example.budgettingapp.database.BudgetDatabaseHelper.GoalEntry INSTANCE = null;
        
        private GoalEntry() {
            super();
        }
    }
}

File: java/com/example/budgettingapp/database/BudgetDatabaseHelper.kt
package com.example.budgettingapp.database

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.provider.BaseColumns
import com.example.budgettingapp.data.Expense
import com.example.budgettingapp.data.Goal
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

class BudgetDatabaseHelper(context: Context) : SQLiteOpenHelper(context, DATABASE_NAME, null, DATABASE_VERSION) {
    
    object ExpenseEntry : BaseColumns {
        const val TABLE_NAME = "expenses"
        const val COLUMN_TITLE = "title"
        const val COLUMN_AMOUNT = "amount"
        const val COLUMN_CATEGORY = "category"
        const val COLUMN_DATE = "date"
        const val COLUMN_START_TIME = "start_time"
        const val COLUMN_END_TIME = "end_time"
        const val COLUMN_DESCRIPTION = "description"
        const val COLUMN_TIMESTAMP = "timestamp"
    }

    object GoalEntry : BaseColumns {
        const val TABLE_NAME = "goals"
        const val COLUMN_TITLE = "title"
        const val COLUMN_TARGET_AMOUNT = "target_amount"
        const val COLUMN_CURRENT_AMOUNT = "current_amount"
        const val COLUMN_DEADLINE = "deadline"
        const val COLUMN_DESCRIPTION = "description"
        const val COLUMN_TIMESTAMP = "timestamp"
    }

    private val SQL_CREATE_ENTRIES = """
        CREATE TABLE ${ExpenseEntry.TABLE_NAME} (
            ${BaseColumns._ID} INTEGER PRIMARY KEY AUTOINCREMENT,
            ${ExpenseEntry.COLUMN_TITLE} TEXT NOT NULL,
            ${ExpenseEntry.COLUMN_AMOUNT} REAL NOT NULL,
            ${ExpenseEntry.COLUMN_CATEGORY} TEXT NOT NULL,
            ${ExpenseEntry.COLUMN_DATE} TEXT NOT NULL,
            ${ExpenseEntry.COLUMN_START_TIME} TEXT NOT NULL,
            ${ExpenseEntry.COLUMN_END_TIME} TEXT NOT NULL,
            ${ExpenseEntry.COLUMN_DESCRIPTION} TEXT,
            ${ExpenseEntry.COLUMN_TIMESTAMP} INTEGER NOT NULL
        );
        CREATE TABLE ${GoalEntry.TABLE_NAME} (
            ${BaseColumns._ID} INTEGER PRIMARY KEY AUTOINCREMENT,
            ${GoalEntry.COLUMN_TITLE} TEXT NOT NULL,
            ${GoalEntry.COLUMN_TARGET_AMOUNT} REAL NOT NULL,
            ${GoalEntry.COLUMN_CURRENT_AMOUNT} REAL NOT NULL,
            ${GoalEntry.COLUMN_DEADLINE} TEXT NOT NULL,
            ${GoalEntry.COLUMN_DESCRIPTION} TEXT,
            ${GoalEntry.COLUMN_TIMESTAMP} INTEGER NOT NULL
        )
    """.trimIndent()

    private val SQL_DELETE_ENTRIES = """
        DROP TABLE IF EXISTS ${ExpenseEntry.TABLE_NAME};
        DROP TABLE IF EXISTS ${GoalEntry.TABLE_NAME}
    """.trimIndent()

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL(SQL_CREATE_ENTRIES)
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        db.execSQL(SQL_DELETE_ENTRIES)
        onCreate(db)
    }

    companion object {
        const val DATABASE_NAME = "budget_database.db"
        const val DATABASE_VERSION = 2
    }

    private val dateFormatter = DateTimeFormatter.ISO_LOCAL_DATE_TIME

    fun insertExpense(expense: Expense): Long {
        val db = writableDatabase
        val values = ContentValues().apply {
            put(ExpenseEntry.COLUMN_TITLE, expense.title)
            put(ExpenseEntry.COLUMN_AMOUNT, expense.amount)
            put(ExpenseEntry.COLUMN_CATEGORY, expense.category)
            put(ExpenseEntry.COLUMN_DATE, dateFormatter.format(expense.date))
            put(ExpenseEntry.COLUMN_START_TIME, dateFormatter.format(expense.startTime))
            put(ExpenseEntry.COLUMN_END_TIME, dateFormatter.format(expense.endTime))
            put(ExpenseEntry.COLUMN_DESCRIPTION, expense.description)
            put(ExpenseEntry.COLUMN_TIMESTAMP, expense.timestamp)
        }
        return db.insert(ExpenseEntry.TABLE_NAME, null, values)
    }

    fun updateExpense(expense: Expense): Int {
        val db = writableDatabase
        val values = ContentValues().apply {
            put(ExpenseEntry.COLUMN_TITLE, expense.title)
            put(ExpenseEntry.COLUMN_AMOUNT, expense.amount)
            put(ExpenseEntry.COLUMN_CATEGORY, expense.category)
            put(ExpenseEntry.COLUMN_DATE, dateFormatter.format(expense.date))
            put(ExpenseEntry.COLUMN_START_TIME, dateFormatter.format(expense.startTime))
            put(ExpenseEntry.COLUMN_END_TIME, dateFormatter.format(expense.endTime))
            put(ExpenseEntry.COLUMN_DESCRIPTION, expense.description)
            put(ExpenseEntry.COLUMN_TIMESTAMP, expense.timestamp)
        }
        return db.update(
            ExpenseEntry.TABLE_NAME,
            values,
            "${BaseColumns._ID} = ?",
            arrayOf(expense.id.toString())
        )
    }



    fun insertGoal(goal: Goal): Long {
        val db = writableDatabase
        val values = ContentValues().apply {
            put(GoalEntry.COLUMN_TITLE, goal.title)
            put(GoalEntry.COLUMN_TARGET_AMOUNT, goal.targetAmount)
            put(GoalEntry.COLUMN_CURRENT_AMOUNT, goal.currentAmount)
            put(GoalEntry.COLUMN_DEADLINE, dateFormatter.format(goal.deadline))
            put(GoalEntry.COLUMN_DESCRIPTION, goal.description)
            put(GoalEntry.COLUMN_TIMESTAMP, goal.timestamp)
        }
        return db.insert(GoalEntry.TABLE_NAME, null, values)
    }

    fun updateGoal(goal: Goal): Int {
        val db = writableDatabase
        val values = ContentValues().apply {
            put(GoalEntry.COLUMN_TITLE, goal.title)
            put(GoalEntry.COLUMN_TARGET_AMOUNT, goal.targetAmount)
            put(GoalEntry.COLUMN_CURRENT_AMOUNT, goal.currentAmount)
            put(GoalEntry.COLUMN_DEADLINE, dateFormatter.format(goal.deadline))
            put(GoalEntry.COLUMN_DESCRIPTION, goal.description)
            put(GoalEntry.COLUMN_TIMESTAMP, goal.timestamp)
        }
        return db.update(
            GoalEntry.TABLE_NAME,
            values,
            "${BaseColumns._ID} = ?",
            arrayOf(goal.id.toString())
        )
    }

    fun getAllGoals(): List<Goal> {
        val db = readableDatabase
        val projection = arrayOf(
            BaseColumns._ID,
            GoalEntry.COLUMN_TITLE,
            GoalEntry.COLUMN_TARGET_AMOUNT,
            GoalEntry.COLUMN_CURRENT_AMOUNT,
            GoalEntry.COLUMN_DEADLINE,
            GoalEntry.COLUMN_DESCRIPTION,
            GoalEntry.COLUMN_TIMESTAMP
        )
        val cursor = db.query(
            GoalEntry.TABLE_NAME,
            projection,
            null,
            null,
            null,
            null,
            null
        )
        val goals = mutableListOf<Goal>()
        with(cursor) {
            while (moveToNext()) {
                val goal = Goal(
                    id = getLong(getColumnIndexOrThrow(BaseColumns._ID)),
                    title = getString(getColumnIndexOrThrow(GoalEntry.COLUMN_TITLE)),
                    targetAmount = getDouble(getColumnIndexOrThrow(GoalEntry.COLUMN_TARGET_AMOUNT)),
                    currentAmount = getDouble(getColumnIndexOrThrow(GoalEntry.COLUMN_CURRENT_AMOUNT)),
                    deadline = LocalDateTime.parse(getString(getColumnIndexOrThrow(GoalEntry.COLUMN_DEADLINE)), dateFormatter),
                    description = getString(getColumnIndexOrThrow(GoalEntry.COLUMN_DESCRIPTION)),
                    timestamp = getLong(getColumnIndexOrThrow(GoalEntry.COLUMN_TIMESTAMP))
                )
                goals.add(goal)
            }
        }
        cursor.close()
        return goals
    }

    fun deleteGoal(goal: Goal): Int {
        val db = writableDatabase
        return db.delete(
            GoalEntry.TABLE_NAME,
            "${BaseColumns._ID} = ?",
            arrayOf(goal.id.toString())
        )
    }



    fun deleteExpense(expense: Expense): Int {
        val db = writableDatabase
        val selection = "${BaseColumns._ID} = ?"
        val selectionArgs = arrayOf(expense.id.toString())
        return db.delete(ExpenseEntry.TABLE_NAME, selection, selectionArgs)
    }

    fun getAllExpenses(): List<Expense> {
        val db = readableDatabase
        val cursor = db.query(
            ExpenseEntry.TABLE_NAME,
            null,
            null,
            null,
            null,
            null,
            "${ExpenseEntry.COLUMN_TIMESTAMP} DESC"
        )
        val expenses = mutableListOf<Expense>()
        with(cursor) {
            while (moveToNext()) {
                val expense = Expense(
                    id = getLong(getColumnIndexOrThrow(BaseColumns._ID)),
                    title = getString(getColumnIndexOrThrow(ExpenseEntry.COLUMN_TITLE)),
                    amount = getDouble(getColumnIndexOrThrow(ExpenseEntry.COLUMN_AMOUNT)),
                    category = getString(getColumnIndexOrThrow(ExpenseEntry.COLUMN_CATEGORY)),
                    date = LocalDateTime.parse(getString(getColumnIndexOrThrow(ExpenseEntry.COLUMN_DATE)), dateFormatter),
                    startTime = LocalDateTime.parse(getString(getColumnIndexOrThrow(ExpenseEntry.COLUMN_START_TIME)), dateFormatter),
                    endTime = LocalDateTime.parse(getString(getColumnIndexOrThrow(ExpenseEntry.COLUMN_END_TIME)), dateFormatter),
                    description = getString(getColumnIndexOrThrow(ExpenseEntry.COLUMN_DESCRIPTION)),
                    timestamp = getLong(getColumnIndexOrThrow(ExpenseEntry.COLUMN_TIMESTAMP))
                )
                expenses.add(expense)
            }
        }
        cursor.close()
        return expenses
    }


    fun getExpenseById(id: Long): Expense? {
        val db = readableDatabase
        val selection = "${BaseColumns._ID} = ?"
        val selectionArgs = arrayOf(id.toString())
        val cursor = db.query(
            ExpenseEntry.TABLE_NAME,
            null,
            selection,
            selectionArgs,
            null,
            null,
            null
        )
        var expense: Expense? = null
        with(cursor) {
            if (moveToNext()) {
                expense = Expense(
                    id = getLong(getColumnIndexOrThrow(BaseColumns._ID)),
                    title = getString(getColumnIndexOrThrow(ExpenseEntry.COLUMN_TITLE)),
                    amount = getDouble(getColumnIndexOrThrow(ExpenseEntry.COLUMN_AMOUNT)),
                    category = getString(getColumnIndexOrThrow(ExpenseEntry.COLUMN_CATEGORY)),
                    date = LocalDateTime.parse(getString(getColumnIndexOrThrow(ExpenseEntry.COLUMN_DATE)), dateFormatter),
                    startTime = LocalDateTime.parse(getString(getColumnIndexOrThrow(ExpenseEntry.COLUMN_START_TIME)), dateFormatter),
                    endTime = LocalDateTime.parse(getString(getColumnIndexOrThrow(ExpenseEntry.COLUMN_END_TIME)), dateFormatter),
                    description = getString(getColumnIndexOrThrow(ExpenseEntry.COLUMN_DESCRIPTION)),
                    timestamp = getLong(getColumnIndexOrThrow(ExpenseEntry.COLUMN_TIMESTAMP))
                )
            }
        }
        cursor.close()
        return expense
    }

    fun getTotalExpenses(startDate: String, endDate: String): Double {
        val db = readableDatabase
        val selection = "${ExpenseEntry.COLUMN_DATE} BETWEEN ? AND ?"
        val selectionArgs = arrayOf(startDate, endDate)
        val cursor = db.rawQuery(
            "SELECT SUM(${ExpenseEntry.COLUMN_AMOUNT}) FROM ${ExpenseEntry.TABLE_NAME} WHERE ${selection}",
            selectionArgs
        )
        var total = 0.0
        with(cursor) {
            if (moveToNext()) {
                total = getDouble(0)
            }
        }
        cursor.close()
        return total
    }

    fun getExpensesByCategory(): List<CategoryTotal> {
        val db = readableDatabase
        val cursor = db.rawQuery(
            "SELECT ${ExpenseEntry.COLUMN_CATEGORY}, SUM(${ExpenseEntry.COLUMN_AMOUNT}) as total FROM ${ExpenseEntry.TABLE_NAME} GROUP BY ${ExpenseEntry.COLUMN_CATEGORY}",
            null
        )
        val totals = mutableListOf<CategoryTotal>()
        with(cursor) {
            while (moveToNext()) {
                totals.add(
                    CategoryTotal(
                        category = getString(getColumnIndexOrThrow(ExpenseEntry.COLUMN_CATEGORY)),
                        total = getDouble(getColumnIndexOrThrow("total"))
                    )
                )
            }
        }
        cursor.close()
        return totals
    }

    data class CategoryTotal(
        val category: String,
        val total: Double
    )
}
